File: competition/First Half 2021/Feb 28. 2021/Solution5692.java
import java.util.*;

class Solution5692 {
	private int[][] copy;

    public double[] getCollisionTimes(int[][] cars) {
    	copy = cars;
    	int size = cars.length;
    	double[] res = new double[size];
    	Stack<Integer> stack = new Stack<>();
    	for(int i = size - 1; i >= 0; i--){
    		while(!stack.empty() && res[stack.peek()] > 0 && (cars[i][1] <= cars[stack.peek()][1] || calTime(i, stack.peek()) > res[stack.peek()])){
    			stack.pop();
    		}
    		if(stack.empty() || cars[i][1] <= cars[stack.peek()][1]){
    			res[i] = -1;
    		}
    		else{
    			res[i] = calTime(i, stack.peek());
    		}
    		stack.push(i);
    	}
    	return res;
    }

	private double calTime(int index1, int index2){
		return (double) (copy[index2][0] - copy[index1][0]) / (double) (copy[index1][1] - copy[index2][1]);
	}
}

File: competition/First Half 2021/Feb 28. 2021/Solution5692.cpp
#include<iostream>
#include<vector>

#include<algorithm>
#include<stack>

using namespace std;

class Solution5692 {
public:

    double calTime(int pos1, int spd1, int pos2, int spd2){
        if(spd1 <= spd2){
            return -1;
        }
        return (double) (pos2 - pos1) / (double) (spd1 - spd2);
    }

    vector<double> getCollisionTimes(vector<vector<int>>& cars) {
        int size = cars.size();
        vector<double> res(size);
        stack<int> stk;
        for(int i = size - 1; i >= 0; i--){
            while(stk.size() > 0 && res[stk.top()] > 0 && (cars[i][1] <= cars[stk.top()][1] || 
            calTime(cars[i][0], cars[i][1], cars[stk.top()][0], cars[stk.top()][1]) > res[stk.top()])){
                stk.pop();
            }
            if(stk.empty() || cars[i][1] <= cars[stk.top()][1]){
                res[i] = -1;
            }
            else{
                res[i] = calTime(cars[i][0], cars[i][1], cars[stk.top()][0], cars[stk.top()][1]);
            }
            stk.push(i);
        }
        return res;
    }
};